tests/test_ec2/test_iam_instance_profile_associations.py

```diff
@@ -0,0 +1,97 @@
+    client = boto3.client("ec2", region_name="us-east-1")
+    conn_ec2 = boto3.resource("ec2", "us-east-1")
+    conn_iam = boto3.resource("iam", "us-east-1")
+    test_instance = conn_ec2.create_instances(ImageId=image_id, MinCount=1, MaxCount=1)
+    test_instance_profile = conn_iam.create_instance_profile(
+        InstanceProfileName="test_profile", Path="/"
+    )
+
+    association = client.associate_iam_instance_profile(
+        IamInstanceProfile={
+            "Arn": test_instance_profile.arn,
+            "Name": test_instance_profile.name,
+        },
+        InstanceId=test_instance[0].id,
+    )
+
+    association["IamInstanceProfileAssociation"]["InstanceId"].should.equal(
+        test_instance[0].id
+    )
+    association["IamInstanceProfileAssociation"]["IamInstanceProfile"][
+        "Arn"
+    ].should.equal(test_instance_profile.arn)
+    association["IamInstanceProfileAssociation"]["State"].should.equal("associating")
+
+    # Describe
+    test_instance = conn_ec2.create_instances(ImageId=image_id, MinCount=1, MaxCount=1)
+    test_instance_profile = conn_iam.create_instance_profile(
+        InstanceProfileName="test_profile2", Path="/"
+    )
+
+    associations = client.describe_iam_instance_profile_associations()
+
+    associations["IamInstanceProfileAssociations"].should.have.length_of(1)
+
+    second_association = client.associate_iam_instance_profile(
+        IamInstanceProfile={
+            "Arn": test_instance_profile.arn,
+            "Name": test_instance_profile.name,
+        },
+        InstanceId=test_instance[0].id,
+    )
+    associations = client.describe_iam_instance_profile_associations()
+    associations["IamInstanceProfileAssociations"].should.have.length_of(2)
+
+    # Disassociate
+    client.disassociate_iam_instance_profile(
+        AssociationId=second_association["IamInstanceProfileAssociation"][
+            "AssociationId"
+        ]
+    )
+    associations = client.describe_iam_instance_profile_associations()
+    associations["IamInstanceProfileAssociations"].should.have.length_of(1)
+
+    # Re-associate
+    new_instance_profile = conn_iam.create_instance_profile(
+        InstanceProfileName="test_profile3", Path="/"
+    )
+    client.replace_iam_instance_profile_association(
+        IamInstanceProfile={
+            "Arn": new_instance_profile.arn,
+            "Name": new_instance_profile.name,
+        },
+        AssociationId=associations["IamInstanceProfileAssociations"][0][
+            "AssociationId"
+        ],
+    )
+    # describe_iam_instance_profile_associations with filter test
+    associations = client.describe_iam_instance_profile_associations(
+        AssociationIds=[
+            associations["IamInstanceProfileAssociations"][0]["AssociationId"],
+        ]
+    )
+    associations["IamInstanceProfileAssociations"][0]["IamInstanceProfile"][
+        "Arn"
+    ].should.equal(new_instance_profile.arn)
+
+    with pytest.raises(ClientError):
```

Comment:
Would be good to assert on the exact error that's begin thrown. Typically we just check the code and message:
```
with pytest.raises() as ex:
    ...
ex.value.response["Error"]["Code"].should.equal(...)
ex.value.response["Error"]["Message"].should.contain(..)
```

Reply:
Invalid tests have been added